### PR TITLE
Fix method name

### PR DESCRIPTION
**Pull request by AlexP11223**

### CODE REVIEW

**Review by XedinUnknown — APPROVED** [review]

**Review by XedinUnknown — CHANGES_REQUESTED** [review]
Needs unit tests, like [here][1].

[1]: https://github.com/wp-oop/wordpress-interface/blob/develop/tests/unit/Plugin/PluginInterfaceTest.php

**Comment by AlexP11223** [comment]
Added tests. Besides the type they also test that we can actually e.g. call `parseBlocks` with a string and it can return an array of blocks.

File: composer.json [hunk 1]
```diff
@@ -26,7 +26,7 @@
   },
   "autoload-dev": {
     "psr-4": {
-      "WpOop\\WordPress\\Test\\Unit\\": "tests/unit"
+      "WpOop\\WordPress\\Test\\": "tests"
```
Comment (XedinUnknown) [inline composer.json]: Well, they are unit tests, and I think it would be nice if unit tests could be found in the same directory in all projects.

File: tests/Stub/Gutenberg/BlockParser.php [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+declare(strict_types=1);
+
+namespace WpOop\WordPress\Test\Stub\Gutenberg;
+
+use WpOop\WordPress\Gutenberg\BlockParserInterface;
+
+class BlockParser implements BlockParserInterface
```
Comment (XedinUnknown) [inline tests/Stub/Gutenberg/BlockParser.php]: That's the thing, see: if you want to create stubs, you are going to add some kind of implementation. I understand that it's just for testing, but why not simply create a mock, like for all other interfaces in all other projects?

File: tests/Unit/Gutenberg/BlockInterfaceTest.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+use WpOop\WordPress\Test\Stub\Gutenberg\Block;
+
+class BlockInterfaceTest extends TestCase
+{
+    public function testUsage()
+    {
+        $innerBlock = new Block('core/quote', [], [], '', []);
+        $subject = new Block(
+            'core/paragraph',
+            ['foo' => 'bar', 'foobar' => 42],
```
Comment (XedinUnknown) [inline tests/Unit/Gutenberg/BlockInterfaceTest.php]: We usually use randomized values where the exact value is not important. This ensures that the implementation is not rigged to pass the tests.

**Review by XedinUnknown — CHANGES_REQUESTED** [review]

File: composer.json (hunk 1 above)
Comment (AlexP11223) [reply composer.json]: They are, it just allows to add other dirs there with this namespace.

File: tests/Stub/Gutenberg/BlockParser.php (hunk 2 above)
Comment (AlexP11223) [reply tests/Stub/Gutenberg/BlockParser.php]: It's simpler, less risk to add a mistake because of misuse of mocking lib etc. In this case there is no logic to mock, it's just a value.

Comment (AlexP11223) [reply tests/Stub/Gutenberg/BlockParser.php]: > like for all other interfaces 

Other tests here test only instantiation. Here :arrow_down: 

>  they also test that we can actually e.g. call `parseBlocks` with a string and it can return an array of blocks.

File: tests/Unit/Gutenberg/BlockInterfaceTest.php (hunk 3 above)
Comment (AlexP11223) [reply tests/Unit/Gutenberg/BlockInterfaceTest.php]: You can rig even with random values, but why would we do that :) 

Here it seems to be fine, the main goal of these tests is to check that the function signatures work as expected, etc., not to test logic.

Comment (XedinUnknown) [reply tests/Unit/Gutenberg/BlockInterfaceTest.php]: How would you rig an implementation if the values that are being tested for are random?

File: tests/Stub/Gutenberg/BlockParser.php (hunk 2 above)
Comment (XedinUnknown) [reply tests/Stub/Gutenberg/BlockParser.php]: I know they test just instantiation. The point in question is: creating extra stubs, VS giving all control to the test itself. I prefer the latter.

File: tests/Unit/Gutenberg/BlockInterfaceTest.php (hunk 3 above)
Comment (AlexP11223) [reply tests/Unit/Gutenberg/BlockInterfaceTest.php]: Depends on what is tested, for example some kind of reflection hack, looking at backtrace, etc. :) Or just returning what was passed without processing, if results are not checked thoroughly enough.

So like if you set a goal to rig tests, you probably can find a way.

File: composer.json (hunk 1 above)
Comment (XedinUnknown) [reply composer.json]: Unit tests differ from e.g. functional or acceptance tests not just by namespace, but by their purpose. That's why they have separate root namespaces. It's that way in all other Dhii/WP-OOP projects.

Comment (XedinUnknown) [reply composer.json]: Then again, like for other similar things: I would love it if you raised this question in a [discussion][] (now enabled), and just do it the same way as everywhere else for now.

[discussion]: https://github.com/wp-oop/wordpress-interface/discussions

Comment (AlexP11223) [reply composer.json]: Nothing changes here about test types or location, everything is the same (except capitalizing the `Unit` dir), it was simply to allow adding other dirs besides `Unit` without duplicating this namespace declaration in `composer.json` and updating `.lock` just to add new dirs (e.g. for functional tests or some helper stuff, for now it was to add `Stub`).

File: tests/Stub/Gutenberg/BlockParser.php (hunk 2 above)
Comment (AlexP11223) [reply tests/Stub/Gutenberg/BlockParser.php]: Creating such basic stubs seems much less complex than duplicating all that via mocking. And more natural, 100% matching how it will behave in real app. PHPUnit mocks seem to have less strict type checks, which is the main purpose of these tests, for example you can call `$subject->getBaseName()` in the plugin test without any errors, while you will get an error if trying to return `null` instead of `string` in stubs.

File: tests/Unit/Gutenberg/BlockInterfaceTest.php (hunk 3 above)
Comment (AlexP11223) [reply tests/Unit/Gutenberg/BlockInterfaceTest.php]: But anyway here adding Faker does not seem to have any benefits because

> the main goal of these tests is to check that the function signatures work as expected, etc., not to test logic.

File: composer.json (hunk 1 above)
Comment (XedinUnknown) [reply composer.json]: I understand the reason. But there is also a reason for doing it the way I have described, like I explained above. Why is it a problem to just declare another root namespace for another kind of tests, until we have a properly argumented discussion and decision about this?

File: tests/Stub/Gutenberg/BlockParser.php (hunk 2 above)
Comment (XedinUnknown) [reply tests/Stub/Gutenberg/BlockParser.php]: What is complicated about `$this->getMockBuilder(BlockParserInterface::class)->getMock()`? Why does it matter what happens when you call `$subject->getBaseName()`, if you are simply testing that the interface can be instantiated corretly, and not invocation of any methods?

File: tests/Unit/Gutenberg/BlockInterfaceTest.php (hunk 3 above)
Comment (XedinUnknown) [reply tests/Unit/Gutenberg/BlockInterfaceTest.php]: In this case, I'm not sure why you need to use any values in the test, or to call any of the methods.

File: tests/Stub/Gutenberg/BlockParser.php (hunk 2 above)
Comment (AlexP11223) [reply tests/Stub/Gutenberg/BlockParser.php]: It is done to test that we do not have any weird mistakes in the function definitions, type-hints, etc., that they work as expected. It seems much more useful than just instantiation (which would not test anything at all in this case, there is no any hierarchy of  interfaces).

File: tests/Unit/Gutenberg/BlockInterfaceTest.php (hunk 3 above)
Comment (AlexP11223) [reply tests/Unit/Gutenberg/BlockInterfaceTest.php]: https://github.com/wp-oop/wordpress-interface/pull/4#discussion_r614076113

File: composer.json (hunk 1 above)
Comment (AlexP11223) [reply composer.json]: :heavy_check_mark:

File: tests/Stub/Gutenberg/BlockParser.php (hunk 2 above)
Comment (XedinUnknown) [reply tests/Stub/Gutenberg/BlockParser.php]: The definitions, including hints, are already provided by the interface, and it therefore seems impossible to make a mistake there. However, the benefit is that the test is fully in control. It's very easy to create the mock, and all the methods area automatically implemented by PHPUnit.

File: tests/Unit/Gutenberg/BlockInterfaceTest.php (hunk 3 above)
Comment (XedinUnknown) [reply tests/Unit/Gutenberg/BlockInterfaceTest.php]: Still don't understand why need to use the values. If you use them, and the implementation is not rigged, then use randomized values to prove that. If you don't need values because you are simply testing instantiation (signature compat etc), then please remove them entirely.

File: tests/Stub/Gutenberg/BlockParser.php (hunk 2 above)
Comment (AlexP11223) [reply tests/Stub/Gutenberg/BlockParser.php]: The idea here is to test that the definitions are sensible and we did not e.g. accidentally add an `object` type hint instead of `mixed` :) So basically this test tries to use it as it was intended, to try typical use cases. And this is also kind of a documentation-like example briefly showing with code how this interface can be used. Also by creating stub we prove that we can actually implement this interface in all PHP versions we support.

This seems to add much more confidence than just testing instantiation.

 Mocking here seems less ideal because of the less strict type checking mentioned above (+ who knows  what other differences from real classes it has), and it would be less convenient to define all methods via mocking (specifying names via strings, etc., instead of just generating class via IDE) which is needed for such tests.

Comment (XedinUnknown) [reply tests/Stub/Gutenberg/BlockParser.php]: > The idea here is to test that the definitions are sensible and we did not e.g. accidentally add an object type hint instead of mixed :)

### TL;DR
This verification is already done by the signatures, and additional testing is redundant.

### Detail
Ah, I see now. Well, my idea for these tests was different: just to check whether signatures are compatible, and to provide a way to verify this should the interfaces change. I do not believe that testing whether a method that returns `object` declares that return type, because this is already a hard requirement of the interface, and the test would simply re-declare that. I do understand a different point of view, however: that design is reflected in a spec, and tests are an implementation of that spec, and therefore they should test whether the code complies with that spec. But this is overkill. IMO this would be acceptable if such tests were created separately and _automatically_. But in that case, you could just generate the interfaces from that same spec, and they would do that validation, so back to square one. Therefore, I would prefer to not do this. I used to do this before, but it really is time-consuming for no good reason IMO. Feel free to create a discussion, though. I would then also post a link in Slack.

Comment (AlexP11223) [reply tests/Stub/Gutenberg/BlockParser.php]: Yeah, since phpcs/psalm already parse and check the code for all PHP versions, these tests are probably not really needed.

The instantiation tests however do not seem to do anything useful too here, but added them for now.

Comment (XedinUnknown) [reply tests/Stub/Gutenberg/BlockParser.php]: Thanks!

**Review by XedinUnknown — APPROVED** [review]
